### ISSUE TITLE
running the standard demo 09-demo-aggregate.yaml against demo env results in and KSMLParseException

### ISSUE THREAD

**Issue by frochse** [Axual/ksml#35]
When i run the 09-demo-aggregate.yaml_ against the local build  I get the following stacktrace:

```
022-04-21T13:39:23,730Z [system] [main] INFO  io.axual.ksml.runner.KSMLRunner - Using backed of type kafka
2022-04-21T13:39:23,735Z [system] [main] INFO  i.a.k.runner.config.KSMLRunnerConfig - Found provider io.axual.ksml.runner.backend.kafka.KafkaBackendProvider for type kafka
2022-04-21T13:39:23,771Z [system] [main] INFO  i.a.k.r.backend.kafka.KafkaBackend - Constructing Kafka Backend
2022-04-21T13:39:23,788Z [system] [main] INFO  i.a.k.g.TopologyGeneratorImpl - Reading KSML from source file(s): [09-demo-aggregate.yaml]
2022-04-21T13:39:24,155Z [system] [main] INFO  io.axual.ksml.user.UserFunction - Registered function 'ksml_pipelines_main_step_1_groupBy_mapper(key:?, value:?) ==> String'
2022-04-21T13:39:29,555Z [system] [main] INFO  io.axual.ksml.user.UserFunction - Registered function 'ksml_pipelines_main_step_3_aggregate_initializer() ==> Long'
2022-04-21T13:39:29,560Z [system] [main] INFO  io.axual.ksml.user.UserFunction - Registered function 'ksml_pipelines_main_step_3_aggregate_aggregator(key:?, value:?, aggregatedValue:?) ==> Long'
Exception in thread "main" io.axual.ksml.exception.KSMLParseException: Error in YAML node 'ksml->pipelines->main->step->3->aggregate': User function definition not found, add 'merger' to specification
        at io.axual.ksml.parser.ContextAwareParser.parseFunction(ContextAwareParser.java:53)
        at io.axual.ksml.parser.ContextAwareParser.parseFunction(ContextAwareParser.java:41)
        at io.axual.ksml.operation.parser.AggregateOperationParser.parse(AggregateOperationParser.java:51)
        at io.axual.ksml.operation.parser.AggregateOperationParser.parse(AggregateOperationParser.java:35)
        at io.axual.ksml.operation.parser.PipelineOperationParser.parse(PipelineOperationParser.java:83)
        at io.axual.ksml.operation.parser.PipelineOperationParser.parse(PipelineOperationParser.java:65)
        at io.axual.ksml.parser.ListParser.parse(ListParser.java:40)
        at io.axual.ksml.definition.parser.PipelineDefinitionParser.parse(PipelineDefinitionParser.java:52)
        at io.axual.ksml.definition.parser.PipelineDefinitionParser.parse(PipelineDefinitionParser.java:43)
        at io.axual.ksml.definition.parser.PipelineDefinitionParser.parse(PipelineDefinitionParser.java:36)
        at io.axual.ksml.parser.MapParser.parse(MapParser.java:40)
        at io.axual.ksml.generator.TopologyGeneratorImpl.generate(TopologyGeneratorImpl.java:156)
        at io.axual.ksml.generator.TopologyGeneratorImpl.create(TopologyGeneratorImpl.java:134)
        at io.axual.ksml.KSMLTopologyGenerator.create(KSMLTopologyGenerator.java:54)
        at io.axual.ksml.runner.backend.kafka.KafkaBackend.<init>(KafkaBackend.java:77)
        at io.axual.ksml.runner.backend.kafka.KafkaBackendProvider.create(KafkaBackendProvider.java:44)
        at io.axual.ksml.runner.config.KSMLRunnerConfig.getConfiguredBackend(KSMLRunnerConfig.java:80)
        at io.axual.ksml.runner.KSMLRunner.main(KSMLRunner.java:54)
```

I did not change a thing to the 09-demo config yaml

**Comment by frochse** [comment]
This only happens if  pull the ksml image from docker-hub "docker pull axual/ksml" if i build this image from the latest source it is ok

**Comment by frochse** [comment]
with the build version from code. The moment the first message is on the stream I get the following 

2022-04-21T15:04:03,962Z [system] [io.ksml.example.processor-a39d17b7-e5a3-4542-a639-a2bee347f6e9-StreamThread-1] ERROR o.apache.kafka.streams.KafkaStreams -
                stream-client [io.ksml.example.processor-a39d17b7-e5a3-4542-a639-a2bee347f6e9] Encountered the following exception during processing and the registered exception handler opted to SHUTDOWN_CLIENT. The streams client is going to shut down now. 
java.lang.NoClassDefFoundError: javax/ws/rs/core/UriBuilder
        at io.confluent.kafka.schemaregistry.client.rest.RestService.getId(RestService.java:656)
        at io.confluent.kafka.schemaregistry.client.rest.RestService.getId(RestService.java:642)
        at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getSchemaByIdFromRegistry(CachedSchemaRegistryClient.java:217)
        at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getSchemaBySubjectAndId(CachedSchemaRegistryClient.java:291)
        at io.confluent.kafka.schemaregistry.client.CachedSchemaRegistryClient.getSchemaById(CachedSchemaRegistryClient.java:276)
        at io.confluent.kafka.serializers.AbstractKafkaAvroDeserializer$DeserializationContext.schemaFromRegistry(AbstractKafkaAvroDeserializer.java:273)
        at io.confluent.kafka.serializers.AbstractKafkaAvroDeserializer.deserialize(AbstractKafkaAvroDeserializer.java:97)
        at io.confluent.kafka.serializers.AbstractKafkaAvroDeserializer.deserialize(AbstractKafkaAvroDeserializer.java:76)
        at io.confluent.kafka.serializers.KafkaAvroDeserializer.deserialize(KafkaAvroDeserializer.java:55)
        at io.axual.ksml.notation.AvroNotation$AvroSerde$2.deserialize(AvroNotation.java:77)
        at org.apache.kafka.common.serialization.Deserializer.deserialize(Deserializer.java:60)
        at org.apache.kafka.streams.processor.internals.SourceNode.deserializeValue(SourceNode.java:58)
        at org.apache.kafka.streams.processor.internals.RecordDeserializer.deserialize(RecordDeserializer.java:66)
        at org.apache.kafka.streams.processor.internals.RecordQueue.updateHead(RecordQueue.java:176)
        at org.apache.kafka.streams.processor.internals.RecordQueue.addRawRecords(RecordQueue.java:112)
        at org.apache.kafka.streams.processor.internals.PartitionGroup.addRawRecords(PartitionGroup.java:185)
        at org.apache.kafka.streams.processor.internals.StreamTask.addRecords(StreamTask.java:957)
        at org.apache.kafka.streams.processor.internals.TaskManager.addRecordsToTasks(TaskManager.java:1009)
        at org.apache.kafka.streams.processor.internals.StreamThread.pollPhase(StreamThread.java:907)
        at org.apache.kafka.streams.processor.internals.StreamThread.runOnce(StreamThread.java:720)
        at org.apache.kafka.streams.processor.internals.StreamThread.runLoop(StreamThread.java:583)
        at org.apache.kafka.streams.processor.internals.StreamThread.run(StreamThread.java:556)
Caused by: java.lang.ClassNotFoundException: javax.ws.rs.core.UriBuilder
        at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:581)
        at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:178)
        at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:522)
        ... 22 common frames omitted

**Comment by tonvanbart** [comment]
Hi Frenk, I checked Docker hub, the axual/ksml image there seems to be some 9 months old. As continuous deployment for this project hasn't been setup properly yet I'd advise against pulling and working with that one for the moment as I'm sure it's outdated.
I would like to ask for clarification on two points, just to be sure:

- when you say you run 09-demo-aggregate.yaml_ against the local build, do you mean you're running it against the docker-compose file provided in the project? If not, how are you running it?
- when you build the latest version from code, how do you do the build? Are you using the supplied Dockerfile-build-runner, or did you install GraalVM locally?

Thanks in advance.

**Comment by tonvanbart** [comment]
Verified that using `Dockerfile-build-runner` to build the runner image, and running against the supplied `docker-compose.yml`, the stack trace above shows up.

**Comment by tonvanbart** [comment]
I think the root cause is that `ksml-query` has a dependency on `jakarta.ws.rs-api:3.0.0` and `ksml` depends on the Confluent schema registry client which transitively depends on `jakarta.ws.rs-api:2.1.6`. Maven tries to resolve the conflict and in the runtime image only the 3.0.0 version is present.
Unfortunately the package naming between the versions is slightly different and what is called `javax.ws.rs.core.UriBuilder` in 2.1.6 is called `jakarta.ws.rs.core.UriBuilder` in 3.0.0.
Tried to resolve by downgrading ksml-query to 2.1.6 and adjusting the imports there, but then I get build errors.

**Comment by frochse** [comment]
Dag Ton,

I am running with the provided docker-compose.
Build is done indeed with the supplied docker-build-runner. 

So no local install, just using the stuff as is. Basically trying all the supplied examples out of the box the get some more experience with KSML. 


Met vriendelijke groet

Frenk ochse
Land in zicht 9
1316 vj Almere
mob:0622513934
***@***.***

**Comment by frochse** [comment]
Hallo Ton,

Dank voor de update

Ik hoor het wel 


Met vriendelijke groet

Frenk ochse
Land in zicht 9
1316 vj Almere
mob:0622513934
***@***.***

**Comment by tonvanbart** [comment]
See also #37

**Comment by tonvanbart** [comment]
Closed as a duplicate of #37 .
